subsys/retention/retention.c

```diff
@@ -0,0 +1,401 @@
+};
+
+static inline void data_retention_lock_take(const struct device *dev)
+{
+#ifdef CONFIG_MULTITHREADING
+	struct retention_data *data = dev->data;
+
+	k_mutex_lock(&data->lock, K_FOREVER);
+#else
+	ARG_UNUSED(dev);
+#endif
+}
+
+static inline void data_retention_lock_release(const struct device *dev)
+{
+#ifdef CONFIG_MULTITHREADING
+	struct retention_data *data = dev->data;
+
+	k_mutex_unlock(&data->lock);
+#else
+	ARG_UNUSED(dev);
+#endif
+}
+
+static int data_retention_checksum(const struct device *dev, uint32_t *output)
+{
+	const struct retention_config *config = dev->config;
+	int rc = -ENOSYS;
+
+	if (config->checksum == CHECKSUM_TYPE_CRC8 ||
+	    config->checksum == CHECKSUM_TYPE_CRC16 ||
+	    config->checksum == CHECKSUM_TYPE_CRC32) {
+		size_t pos = config->offset + config->prefix_len;
+		size_t end = config->offset + config->size - config->checksum;
+		uint8_t buffer[CONFIG_RETENTION_BUFFER_SIZE];
+
+		*output = 0;
+
+		while (pos < end) {
+			uint8_t read_size = MIN((end - pos), sizeof(buffer));
+
+			rc = retained_mem_read(config->parent, pos, buffer, read_size);
+
+			if (rc < 0) {
+				goto finish;
+			}
+
+			if (config->checksum == CHECKSUM_TYPE_CRC8) {
+				*output = (uint32_t)crc8(buffer, read_size, 0x12,
+							 (uint8_t)*output, false);
+			} else if (config->checksum == CHECKSUM_TYPE_CRC16) {
+				*output = (uint32_t)crc16_itu_t((uint16_t)*output,
+								buffer, read_size);
+			} else if (config->checksum == CHECKSUM_TYPE_CRC32) {
+				*output = crc32_ieee_update(*output, buffer, read_size);
+			}
+
+			pos += read_size;
+		}
+	}
+
+finish:
+	return rc;
+}
+
+static int data_retention_init(const struct device *dev)
+{
+	const struct retention_config *config = dev->config;
+#ifdef CONFIG_MULTITHREADING
+	struct retention_data *data = dev->data;
+#endif
+	ssize_t area_size;
+
+	if (!device_is_ready(config->parent)) {
+		LOG_ERR("Parent device is not ready");
+		return -ENODEV;
+	}
+
+	/* Ensure backend has a large enough storage area for the requirements of
+	 * this retention area
+	 */
+	area_size = retained_mem_size(config->parent);
+
+	if (area_size < 0) {
+		LOG_ERR("Parent initialisation failure: %d", area_size);
+		return area_size;
+	}
+
+	if ((config->offset + config->size) > area_size) {
+		/* Backend storage is insufficient */
+		LOG_ERR("Underlying area size is insufficient, requires: 0x%x, has: 0x%x",
+			(config->offset + config->size), area_size);
+		return -EINVAL;
+	}
+
+#ifdef CONFIG_MULTITHREADING
+	k_mutex_init(&data->lock);
+#endif
+
+	return 0;
+}
+
+ssize_t data_retention_size(const struct device *dev)
+{
+	const struct retention_config *config = dev->config;
+
+	return (config->size - config->reserved_size);
+}
+
+int data_retention_is_valid(const struct device *dev)
+{
+	const struct retention_config *config = dev->config;
+	struct retention_data *data = dev->data;
+	int rc = 0;
+	uint8_t buffer[CONFIG_RETENTION_BUFFER_SIZE];
+	off_t pos;
+
+	data_retention_lock_take(dev);
```

Comment:
quick question, are the locking mechanisms really necessary containing the underlying subsys `retained_mem` has its own synchronization primitives ?

Reply:
If there are 2 threads using this system and there are locks on the retained memory but not on the retention subsystem, one thread can call a write function which starts writing data to the retained memory, say an interrupt occurs and now the other thread writes to the same retention area - what will the data look like once both functions have returned? I do take your point though that there should be a way to disable using mutexes if the user so wishes not to (at their own risk) so can add a Kconfig to both retained memory and retention to disable mutex support

Reply:
ok gotcha, I didn't figure out that the `data_retention_write` was making several calls to `retained_mem_write` in that case  you could have partial writes which would be annoying error cases to handle.